File: 15-3sum/15-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res=[]
        
        if len(nums)<3:
            return []
        
        for i in range(len(nums)-1):
            if i!=0 and nums[i]==nums[i-1]: continue
            L,R=i+1,len(nums)-1
            while L<R:
                curSum=nums[L]+nums[R]+nums[i] 
                if curSum==0:
                    res.append([nums[L],nums[R],nums[i]])
                    L+=1
                    R-=1
                    while L<R and nums[L]==nums[L-1]:
                        L+=1
                    while L<R and nums[R]==[R+1]:
                        R-=1
                elif curSum<0:
                    L+=1
                else:
                    R-=1
        return res
                
            
        

File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
           //base case
        List<List<Integer>> res= new ArrayList<>();
        if(nums.length<3)return res;
        Arrays.sort(nums);
        int n=nums.length;

        for (int i = 0; i < n; i++) {
            if(i!=0 && nums[i]==nums[i-1])continue;

            int L=i+1,R=n-1;
            while (L<R){
                int curSum=nums[i]+nums[L]+nums[R];
                if(curSum==0){
                    res.add(Arrays.asList(nums[i],nums[L],nums[R]));
                    L++;
                    R--;
                    //look for unique elements
                    while (L<R && nums[L]==nums[L-1])L++;
                    while (L<R && nums[R]==nums[R+1])R--;
                } else if (curSum<0) {
                    L++;

                }else {
                    R--;
                }
            }
        }

      return res;
    }
}